sql/src/main/java/org/apache/druid/sql/calcite/rule/RewriteFirstValueLastValueRule.java

```diff
@@ -0,0 +1,125 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.sql.calcite.rule;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.rules.SubstitutionRule;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexOver;
+import org.apache.calcite.rex.RexShuttle;
+import org.apache.calcite.rex.RexWindow;
+import org.apache.calcite.rex.RexWindowBound;
+import org.apache.calcite.sql.SqlAggFunction;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+
+import java.math.BigDecimal;
+import java.util.List;
+
+public class RewriteFirstValueLastValueRule extends RelOptRule implements SubstitutionRule
+{
+  public RewriteFirstValueLastValueRule()
+  {
+    super(operand(RelNode.class, any()));
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call)
+  {
+    final RelNode oldNode = call.rel(0);
+    final RewriteShuttle shuttle = new RewriteShuttle(oldNode.getCluster().getRexBuilder());
+    final RelNode newNode = oldNode.accept(shuttle);
+
+    // noinspection ObjectEquality
+    if (newNode != oldNode) {
+      call.transformTo(newNode);
+      call.getPlanner().prune(oldNode);
+    }
+  }
+
+  private static class RewriteShuttle extends RexShuttle
+  {
+    private final RexBuilder rexBuilder;
+
+    public RewriteShuttle(RexBuilder rexBuilder)
+    {
+      this.rexBuilder = rexBuilder;
+    }
+
+    @Override
+    public RexNode visitOver(RexOver over)
+    {
+      SqlOperator operator = over.getOperator();
+      RexWindow window = over.getWindow();
+      RexWindowBound upperBound = window.getUpperBound();
+      RexWindowBound lowerBound = window.getLowerBound();
+
+      if (window.orderKeys.size() > 0) {
+        if (operator.getKind() == SqlKind.LAST_VALUE && !upperBound.isUnbounded()) {
+          if (upperBound.isCurrentRow()) {
+            return rewriteToReferenceCurrentRow(over);
+          }
+        }
+        if (operator.getKind() == SqlKind.FIRST_VALUE && !lowerBound.isUnbounded()) {
+          if (lowerBound.isCurrentRow()) {
+            return rewriteToReferenceCurrentRow(over);
+          }
+        }
+      }
+      return super.visitOver(over);
+    }
+
+    private RexNode rewriteToReferenceCurrentRow(RexOver over)
+    {
+      // could remove `last_value( x ) over ( .... order by y )`
+      // best would be to: return over.getOperands().get(0);
+      // however that make some queries too good
+      return makeOver(
+          over,
+          over.getWindow(),
+          SqlStdOperatorTable.LAG,
```

Comment:
If the operator.getKind() is LAST_VALUE we transform to LAG(x,0) but if the operator.getKind() is FIRST_VALUE do we need to transform to LEAD(x,0) ? Probably a comment can help here